RepoDb/RepoDb/Enumerations/Conjunction.cs

```diff
@@ -5,15 +5,15 @@ namespace RepoDb.Enumerations
     /// <summary>
     /// An enumeration used to define a conjuction for the query grouping. This enumeration is used at <see cref="QueryGroup"/> object.
     /// </summary>
-    public enum Conjunction : short
+    public enum Conjunction
     {
         /// <summary>
         /// The (AND) conjunction.
         /// </summary>
-        [Text("AND")] And = 1,
+        [Text("AND")] And = 446274343,
```

Comment:
@calumchisholm - It will definitely improve the performance as discussed on our separate thread.  Btw, how did you generated the 'value' here, it looks like an actual 'hashCode'.

Would it not be possible to just let the value 'defaulted to constant value 1,2,3,4...' as standard(ed) by Microsoft? And in the 'GetHashCode()' method of the '<Operation>Request' classes, we will just cast it to the '(int)Conjunction, (int)Order, (int)Operation'.

May I request you do a change here by simply putting back the old values '1,2,3,4...'? Or else, please revert back and issue the reason why we should do such hashcode value assignment over normal value assignment.

Reply:
@mikependon - the numbers aren't true hash values, but arbitrary random numbers generated using `System.Random`. The intention being that they're non-consecutive, unique across all 3 enums and (I assume) well distributed. The reason we can't use consecutive numbers is that they dramatically increase the likelihood of collisions.

For example, if we used consecutive runs of numbers values for the `Order` and `Conjunction` enums, the following queries would collide:

`SELECT * FROM Users WHERE (x AND y) ORDER BY z DESC`
`SELECT * FROM Users WHERE (x OR y) ORDER BY z ASC`

I believe this is probably also an issue in the current version of the code (though somewhat hidden), where
`GetHashCode(Order.ASC) == GetHashCode(Conjunction.AND) == GetHashCode(1)`

I think there are also some broader problems relating to the use of hashing in RepoDb. For example, hashed values are currently summed, causing the distribution to become skewed - longer queries generate progressively higher hash values. XORing hash values together is more valid/accepted approach, though still not perfect. Using XOR may also bring small efficiency gains, as it doesn't require any carry / overflow checking (currently unhandled?).

More seriously though, the nature of hashing means that:
 * Two hash values which are unequal **guarantees** that the objects they represent are different.
 * Two hash values which are equal **doesn't** guarantee that their objects are equal (due to the risk of collisions), only that they *may* be.

For this reason, hash values should never be used to test equality (only *in*equality). This isn't some abstract risk either - the probability of a collision within a 32-bit space is [surprisingly high](https://preshing.com/20110504/hash-collision-probabilities/), and obviously increases with the number of hashes calculated.

I'd be interested to see how this problem is addressed in similar situations. Perhaps I should raise a separate issue?

Reply:
Hi @calumchisholm ,

First, thank you for digging into this area. I am very flattered! Just to let you know that this is one of the recipes as explained by me in this article (http://codesdirectory.blogspot.com/2018/09/repodb-net-lightweight-orm-library.html).

You are actually understanding the recipe of RepoDb on how I made it much more faster (probably) over the other. As a result of caching everything in the memory for the SQL Statement generated via IConcurrentDictionary (thread-safe).

It is already proven officially by RawDataAccessBencher that RepoDb performance and memory-efficiencies are ahead over the other ORMs. But this does not mean that there is NO window of improvements on here.

Okay, the collision is one of the ealiest problem I solved on this combination. Based on the way how I generated the simple 'GetHashCode()', the collision will never ever happen based on the combination.

Below is the way how I overrided the equality comparers via overriding the 'GetHashCode()', and if anyone says this is not the right way to do (or somehow anti-pattern), then so be it. The results matter (speed, performance and the efficiencies of RepoDb matters for me).

- Field.GetHashCode() = RepoDb.Field.Name.GetHashCode()
- Parameter.GetHashCode() = RepoDb.Parameter.Name.GetHashCode()
- OrderField.GetHashCode() = RepoDb.OrderField.Name.GetHashCode() + RepoDb.Enumerations.Order.GetHashCode()
- QueryField.GetHashCode() = RepoDb.Field.GetHashCode() + RepoDb.Operation.GetHashCode() + RepoDb.Parameter.GetHashCode()
	For Operation.Equal (IS NULL): + 128
	For Operation.NotEqual (IS NOT NULL): + 256
	Above are both constant values to avoid the collisions.
	For Operation.<In|NotIn> (Array Params like IN (@FieldA_1, @FieldA_2, @FieldA_3): + Array.GetHashCode()
- QueryGroup.GetHashCode() = SUM((foreach)QueryFields.GetHashCode()) + SUM((foreach)QueryGroups.GetHashCode()) + RepoDb.Enumerations.Conjunction.GetHashCode() + RepoDb.QueryGroup.IsNot.GetHashCode()
	
Your PR is major request on the core design of the library performance and efficiencies and I need to be very careful on here.

The collision problems were solved by the 'Integration Tests' that I have written. Where I write the test in the following pseudocodes.

	- Test the Field equality by Name
	- Test the Field inequality by Name
	- Test the Field equality by Name (improper)
	- Test the OrderField equality by Name (ASC)
	- Test the OrderField equality by Name (DESC)
	- Test the OrderField equality by Name (improper, ASC)
	- Test the OrderField equality by Name (improper, DESC)
	- And other hundred Unit Tests and Integration Tets....

They were also covered by the 'Unit Tests', which can be seen in both 'Tests' project.

By setting the values of (1, 2, 3, 4 ....) in this Enumerations would not really affect the collision, and thinking to the fact that collision could be a major problem in the future, then I must to say "do not think too much on it, it is already covered in the first 'GetHashCode()' coding development".

--------------

P.S:

If we are not really satisfied with it, we can simply add a simple multiplier on each object, where we can simply introduce the following.

	- QueryField - x 3
	- QueryGroup - x 4
	- Field - x 5
	- OrderField - x 6
	- And so forth...

This is just to guarantee that the collision will never ever happen here, and again, covered by the tests.

In your sample, here:

     SELECT * FROM Users WHERE (x AND y) ORDER BY z DESC
     SELECT * FROM Users WHERE (x OR y) ORDER BY z ASC

They will NEVER EVER collide because the Conjunction is different (AND and OR) will generate a different hashcodes.

--------------

I am open to discuss this more further because I am very keen of making this library the craziest, so please allow yourself to push-back on here if you have an idea that is not considered here.

And again, you rocks by digging on RepoDb big secret.

Reply:
Opps! But, since you included the ORDER BY (Order), then you are correct that they will collide on here if we are to use the default values (1,2,3,4...), as we are abandoning the ‘GetHashCode()’ calls of the enums.

I am keen on putting the (1,2,3,4...) values and add a multiplier, over the big random generated values. In this way, other developers would somehow understand the code easily and is still on standard way of implementing the enums.

So, I prefer this in QueryField (and so do to others like QueryGroup).

     QueryField.GetHashCode() = Field.GetHashCode() + ((int)Operation * OPERATION_MULTIPLIER) + Parameter.GetHashCode().

Atleast with this, the control is on the class object and not in the enum values. IMHO

Reply:
Merging this to the main branch, kindly create a new PR if you wish to reconsider the multiplier approach.